pallets/gear/src/queue.rs

```diff
@@ -61,49 +51,96 @@ where
             block_config,
             GasAllowanceOf::<T>::get(),
             dispatch.into_incoming(gas_limit),
-            program_id,
+            destination_id,
         ) {
             Ok(dispatch) => dispatch,
-            Err(journal) => return Ok(journal),
+            Err(journal) => return journal,
+        };
+
+        let Some(program) =
+            ProgramStorageOf::<T>::get_program(destination_id).and_then(|p| match p {
+                Program::Active(p) => Some(p),
+                _ => None,
+            })
+        else {
+            log::trace!("Message is sent to non-active program {destination_id:?}");
+            return core_processor::process_non_executable(precharged_dispatch, destination_id);
         };
 
-        let (precharged_dispatch, actor_data) =
-            get_actor_data(precharged_dispatch).map_err(QueueStepError::ActorData)?;
+        if program.state == ProgramState::Initialized && dispatch_kind == DispatchKind::Init {
+            // Panic is impossible, because gear protocol does not provide functionality
+            // to send second init message to any already existing program.
+            unreachable!(
+                "Init message {dispatch_id:?} is sent to already initialized program {destination_id:?}"
+            );
+        }
+
+        // If the destination program is uninitialized, then we allow
+        // to process message, if it's a reply or init message.
+        // Otherwise, we append the message to the waiting init message list.
+        if matches!(program.state, ProgramState::Uninitialized { message_id }
+            if message_id != dispatch_id && dispatch_kind != DispatchKind::Reply)
+        {
+            if dispatch_kind == DispatchKind::Init {
+                // Panic is impossible, because gear protocol does not provide functionality
+                // to send second init message to any existing program.
+                unreachable!(
+                    "Init message {dispatch_id:?} is not the first init message to the program {destination_id:?}"
+                );
+            }
+
+            let (dispatch, gas, _) = precharged_dispatch.into_parts();
+            return vec![
+                JournalNote::GasBurned {
+                    message_id: dispatch.id(),
+                    amount: gas.burned(),
+                },
+                JournalNote::WaitingInitMessage {
```

Comment:
Looks like this is the only place where this journal note is used. Seems to be very ad-hoc-ish to me because journal notes expected to be created on `core-processor` side and handled on `pallet` side

Reply:
So, I can make fn process_waiting_init() -> Vec<JournalNote> in core-processor, but this is not make scence

Reply:
The same problem like I mentioned here :/

https://github.com/gear-tech/gear/pull/3714#discussion_r1480781414